### ISSUE TITLE
Different kernel obtained given equations in paper

### ISSUE THREAD

**Issue by dbadrian** [HoseinHasani/SM-CNN#1]
Dear authors,

I was hoping to run some experiments with the code - is there a release planned anytime soon?
I noticed that using the equations (1,2) and the numbers provided for sigma_i and sigma_e, that I obtained a different kernel, than the one provided in Figure 1b?
Is the one in figure 1b just representative and not the actual one used?

Best
David

**Comment by HoseinHasani** [comment]
Hi @dbadrian 
Sorry for the late response.

- Unfortunately, my schedule is full and I'm so busy these days. I plan to rewrite all of the codes in TensorFlow v2 and then release them. 

- You're right. The kernel from Figure 1b is slightly different from that explicitly derived from setting sigma_i and sigma_e to 1.4 and 1.2. In fact, we provided some explanation about this in the camera-ready version but we finally decided to omit them because this part was not directly asked by reviewers and only minor changes were allowed in the final version.
We found that the excitatory-inhibitory balance is important and our experiments indicated that an SM kernel with a near-zero average is more effective. This balance is also considered in other biological models like [this](https://www.cell.com/neuron/pdf/S0896-6273(19)30377-0.pdf). For more information please read [this Scholarpedia article](http://www.scholarpedia.org/article/Balance_of_excitation_and_inhibition) .
One way to guarantee near-zero average for small DoG kernels is to reflect the out of the border tails of the Gaussian functions into the excitatory and inhibitory kernels. For example, this can be done by setting `mode='reflect'` in `scipy.ndimage.gaussian_filter`.

**Comment by dbadrian** [comment]
Dear @HoseinHasani,

thank you for getting back, very much appreciated, that resolves the ambiguity.

In case the new project might not be released soonish, I was wondering if you could give me a pointer w.r.t. implementation used, since its not clear to me from the paper/supplement:
Specifically, I was wondering, for the SM-ResNet18 (supplements): I assume you used the common architecture proposed in https://arxiv.org/pdf/1512.03385.pdf for the BasicBlocks:

1. 3x3 conv
2. batch_norm
3. relu
4. 3x3 conv
5. batch_norm

In between what numbers did you place the SM. Just 3 and 4 of the very first block?

Thank you very much!

**Comment by HoseinHasani** [comment]
Yes, we applied SM kernel on half of the activation maps between 3 and 4. But you should avoid using batch norm in level 2 (bach norms from higher layers can influence performance too). Our implementation was based on "Pre-activation ResNet":
[https://arxiv.org/pdf/1603.05027](https://arxiv.org/pdf/1603.05027)

You're welcome ;)

**Comment by dbadrian** [comment]
Oh, I see - some important detail ;).

Thanks for the input!
